pkg/manifest/os.go

```diff
@@ -737,6 +739,36 @@ func (p *OS) serialize() osbuild.Pipeline {
 		pipeline.AddStages(osbuild.GenFileNodesStages(p.Files)...)
 	}
 
+	// write modularity related configuration files
+	if len(p.moduleSpecs) > 0 {
+		pipeline.AddStages(osbuild.GenDNFModuleConfigStages(p.moduleSpecs)...)
+
+		var failsafeFiles []*fsnode.File
+
+		// the failsafe file is a blob of YAML returned directly from the depsolver,
+		// we write them as 'normal files' without a special stage
+		for _, module := range p.moduleSpecs {
+			moduleFailsafeFile, err := fsnode.NewFile(module.FailsafeFile.Path, nil, nil, nil, []byte(module.FailsafeFile.Data))
+
+			if err != nil {
+				panic("failed to create module failsafe file")
+			}
+
+			failsafeFiles = append(failsafeFiles, moduleFailsafeFile)
+		}
+
+		failsafeDir, err := fsnode.NewDirectory("/var/lib/dnf/modulefailsafe", nil, nil, nil, true)
+
+		if err != nil {
+			panic("failed to create module failsafe directory")
+		}
+
+		pipeline.AddStages(osbuild.GenDirectoryNodesStages([]*fsnode.Directory{failsafeDir})...)
+		pipeline.AddStages(osbuild.GenFileNodesStages(failsafeFiles)...)
+
+		p.Files = append(p.Files, failsafeFiles...)
```

Comment:
Note for another time (future refactoring): We could pull the failsafe file handling out into a separate function, either here or in osbuild as a helper function in `osbuild/dnf_module_config_stage.go`.  The work to generate the files is localised here so it's not a big issue, there's no duplication, but it would make the OS pipeline generation easier to read.

Reply:
+1 I had a very similar comment, which has already been resolved.